### ISSUE TITLE
Invocation of Java chaincode times out

### ISSUE THREAD

**Issue by bzp99** [ftsrg/blockchain-benchmarks-tpcc#8]
After ~resolving~ working around #5, this is the next roadblock. After starting a clean fablo network, and installing the CC, invoking `InitEntries` behaves as such:

```
$ http -v -A bearer -a $token http://localhost:8801/invoke/my-channel1/tpcc method=InitEntries args:='[]'
POST /invoke/my-channel1/tpcc HTTP/1.1
Accept: application/json, */*;q=0.5
Accept-Encoding: gzip, deflate
Authorization: Bearer 12bd3ca0-1a58-11ee-8f78-f1a42b6481c8-admin
Connection: keep-alive
Content-Length: 37
Content-Type: application/json
Host: localhost:8801
User-Agent: HTTPie/3.2.1

{
    "args": [],
    "method": "InitEntries"
}


HTTP/1.1 400 Bad Request
Access-Control-Allow-Origin: *
Cache-Control: no-store
Connection: keep-alive
Content-Length: 121
Content-Type: application/json; charset=utf-8
Date: Tue, 04 Jul 2023 10:47:17 GMT
ETag: W/"79-fcvDXAnvjjKKtv/YS7bVy48Vnww"
Keep-Alive: timeout=5
X-Powered-By: Express

{
    "message": "No valid responses from any peers. Errors:\n    peer=undefined, status=grpc, message=Endorsement has failed"
}
```

The cause of the error (timeout) is clearer from the peer logs:

```
2023-07-04 10:47:17.690 UTC 0063 INFO [endorser] callChaincode -> finished chaincode: tpcc duration: 30010ms channel=my-channel1 txID=4f9806d4
2023-07-04 10:47:17.690 UTC 0064 ERRO [endorser] simulateProposal -> failed to invoke chaincode tpcc, error: timeout expired while executing transaction
github.com/hyperledger/fabric/core/chaincode.(*Handler).Execute
        /go/src/github.com/hyperledger/fabric/core/chaincode/handler.go:1188
github.com/hyperledger/fabric/core/chaincode.(*ChaincodeSupport).execute
        /go/src/github.com/hyperledger/fabric/core/chaincode/chaincode_support.go:278
github.com/hyperledger/fabric/core/chaincode.(*ChaincodeSupport).Invoke
        /go/src/github.com/hyperledger/fabric/core/chaincode/chaincode_support.go:208
github.com/hyperledger/fabric/core/chaincode.(*ChaincodeSupport).Execute
        /go/src/github.com/hyperledger/fabric/core/chaincode/chaincode_support.go:161
github.com/hyperledger/fabric/core/endorser.(*SupportImpl).Execute
        /go/src/github.com/hyperledger/fabric/core/endorser/support.go:126
github.com/hyperledger/fabric/core/endorser.(*Endorser).callChaincode
        /go/src/github.com/hyperledger/fabric/core/endorser/endorser.go:120
github.com/hyperledger/fabric/core/endorser.(*Endorser).simulateProposal
        /go/src/github.com/hyperledger/fabric/core/endorser/endorser.go:187
github.com/hyperledger/fabric/core/endorser.(*Endorser).ProcessProposalSuccessfullyOrError
        /go/src/github.com/hyperledger/fabric/core/endorser/endorser.go:409
github.com/hyperledger/fabric/core/endorser.(*Endorser).ProcessProposal
        /go/src/github.com/hyperledger/fabric/core/endorser/endorser.go:350
github.com/hyperledger/fabric/core/handlers/auth/filter.(*expirationCheckFilter).ProcessProposal
        /go/src/github.com/hyperledger/fabric/core/handlers/auth/filter/expiration.go:61
github.com/hyperledger/fabric/core/handlers/auth/filter.(*filter).ProcessProposal
        /go/src/github.com/hyperledger/fabric/core/handlers/auth/filter/filter.go:32
github.com/hyperledger/fabric-protos-go/peer._Endorser_ProcessProposal_Handler.func1
        /go/src/github.com/hyperledger/fabric/vendor/github.com/hyperledger/fabric-protos-go/peer/peer.pb.go:107
github.com/hyperledger/fabric/internal/peer/node.unaryGrpcLimiter.func1
        /go/src/github.com/hyperledger/fabric/internal/peer/node/grpc_limiters.go:56
github.com/grpc-ecosystem/go-grpc-middleware.ChainUnaryServer.func1.1.1
        /go/src/github.com/hyperledger/fabric/vendor/github.com/grpc-ecosystem/go-grpc-middleware/chain.go:25
github.com/hyperledger/fabric/common/grpclogging.UnaryServerInterceptor.func1
        /go/src/github.com/hyperledger/fabric/common/grpclogging/server.go:92
github.com/grpc-ecosystem/go-grpc-middleware.ChainUnaryServer.func1.1.1
        /go/src/github.com/hyperledger/fabric/vendor/github.com/grpc-ecosystem/go-grpc-middleware/chain.go:25
github.com/hyperledger/fabric/common/grpcmetrics.UnaryServerInterceptor.func1
        /go/src/github.com/hyperledger/fabric/common/grpcmetrics/interceptor.go:31
github.com/grpc-ecosystem/go-grpc-middleware.ChainUnaryServer.func1.1.1
        /go/src/github.com/hyperledger/fabric/vendor/github.com/grpc-ecosystem/go-grpc-middleware/chain.go:25
github.com/grpc-ecosystem/go-grpc-middleware.ChainUnaryServer.func1
        /go/src/github.com/hyperledger/fabric/vendor/github.com/grpc-ecosystem/go-grpc-middleware/chain.go:34
github.com/hyperledger/fabric-protos-go/peer._Endorser_ProcessProposal_Handler
        /go/src/github.com/hyperledger/fabric/vendor/github.com/hyperledger/fabric-protos-go/peer/peer.pb.go:109
google.golang.org/grpc.(*Server).processUnaryRPC
        /go/src/github.com/hyperledger/fabric/vendor/google.golang.org/grpc/server.go:1180
google.golang.org/grpc.(*Server).handleStream
        /go/src/github.com/hyperledger/fabric/vendor/google.golang.org/grpc/server.go:1503
google.golang.org/grpc.(*Server).serveStreams.func1.2
        /go/src/github.com/hyperledger/fabric/vendor/google.golang.org/grpc/server.go:843
runtime.goexit
        /usr/local/go/src/runtime/asm_amd64.s:1581
error sending
failed to execute transaction 4f9806d4cf425c21302a71087f24ddbbd34e231280ac8526a4c182b32ecf20cc
github.com/hyperledger/fabric/core/chaincode.processChaincodeExecutionResult
        /go/src/github.com/hyperledger/fabric/core/chaincode/chaincode_support.go:167
github.com/hyperledger/fabric/core/chaincode.(*ChaincodeSupport).Execute
        /go/src/github.com/hyperledger/fabric/core/chaincode/chaincode_support.go:162
github.com/hyperledger/fabric/core/endorser.(*SupportImpl).Execute
        /go/src/github.com/hyperledger/fabric/core/endorser/support.go:126
github.com/hyperledger/fabric/core/endorser.(*Endorser).callChaincode
        /go/src/github.com/hyperledger/fabric/core/endorser/endorser.go:120
github.com/hyperledger/fabric/core/endorser.(*Endorser).simulateProposal
        /go/src/github.com/hyperledger/fabric/core/endorser/endorser.go:187
github.com/hyperledger/fabric/core/endorser.(*Endorser).ProcessProposalSuccessfullyOrError
        /go/src/github.com/hyperledger/fabric/core/endorser/endorser.go:409
github.com/hyperledger/fabric/core/endorser.(*Endorser).ProcessProposal
        /go/src/github.com/hyperledger/fabric/core/endorser/endorser.go:350
github.com/hyperledger/fabric/core/handlers/auth/filter.(*expirationCheckFilter).ProcessProposal
        /go/src/github.com/hyperledger/fabric/core/handlers/auth/filter/expiration.go:61
github.com/hyperledger/fabric/core/handlers/auth/filter.(*filter).ProcessProposal
        /go/src/github.com/hyperledger/fabric/core/handlers/auth/filter/filter.go:32
github.com/hyperledger/fabric-protos-go/peer._Endorser_ProcessProposal_Handler.func1
        /go/src/github.com/hyperledger/fabric/vendor/github.com/hyperledger/fabric-protos-go/peer/peer.pb.go:107
github.com/hyperledger/fabric/internal/peer/node.unaryGrpcLimiter.func1
        /go/src/github.com/hyperledger/fabric/internal/peer/node/grpc_limiters.go:56
github.com/grpc-ecosystem/go-grpc-middleware.ChainUnaryServer.func1.1.1
        /go/src/github.com/hyperledger/fabric/vendor/github.com/grpc-ecosystem/go-grpc-middleware/chain.go:25
github.com/hyperledger/fabric/common/grpclogging.UnaryServerInterceptor.func1
        /go/src/github.com/hyperledger/fabric/common/grpclogging/server.go:92
github.com/grpc-ecosystem/go-grpc-middleware.ChainUnaryServer.func1.1.1
        /go/src/github.com/hyperledger/fabric/vendor/github.com/grpc-ecosystem/go-grpc-middleware/chain.go:25
github.com/hyperledger/fabric/common/grpcmetrics.UnaryServerInterceptor.func1
        /go/src/github.com/hyperledger/fabric/common/grpcmetrics/interceptor.go:31
github.com/grpc-ecosystem/go-grpc-middleware.ChainUnaryServer.func1.1.1
        /go/src/github.com/hyperledger/fabric/vendor/github.com/grpc-ecosystem/go-grpc-middleware/chain.go:25
github.com/grpc-ecosystem/go-grpc-middleware.ChainUnaryServer.func1
        /go/src/github.com/hyperledger/fabric/vendor/github.com/grpc-ecosystem/go-grpc-middleware/chain.go:34
github.com/hyperledger/fabric-protos-go/peer._Endorser_ProcessProposal_Handler
        /go/src/github.com/hyperledger/fabric/vendor/github.com/hyperledger/fabric-protos-go/peer/peer.pb.go:109
google.golang.org/grpc.(*Server).processUnaryRPC
        /go/src/github.com/hyperledger/fabric/vendor/google.golang.org/grpc/server.go:1180
google.golang.org/grpc.(*Server).handleStream
        /go/src/github.com/hyperledger/fabric/vendor/google.golang.org/grpc/server.go:1503
google.golang.org/grpc.(*Server).serveStreams.func1.2
        /go/src/github.com/hyperledger/fabric/vendor/google.golang.org/grpc/server.go:843
runtime.goexit
        /usr/local/go/src/runtime/asm_amd64.s:1581 channel=my-channel1 txID=4f9806d4
2023-07-04 10:47:17.690 UTC 0065 WARN [endorser] ProcessProposal -> Failed to invoke chaincode channel=my-channel1 chaincode=tpcc error="error in simulation: failed to execute transaction 4f9806d4cf425c21302a71087f24ddbbd34e231280ac8526a4c1
82b32ecf20cc: error sending: timeout expired while executing transaction"
2023-07-04 10:47:17.690 UTC 0066 INFO [comm.grpc.server] 1 -> unary call completed grpc.service=protos.Endorser grpc.method=ProcessProposal grpc.peer_address=192.168.192.8:56666 grpc.code=OK grpc.call_duration=30.01s
```

Strangely, there seems to be no output from the chaincode container. This makes me think that this is a connection issue between the Fabric network and the external CC. Digging deeper…

**Comment by bzp99** [comment]
FWIW, it really does seem that the invocation does not even reach the CC container:

```
$ docker run --rm -it --network [...]_basic tcpdump tcpdump port 9999
tcpdump: verbose output suppressed, use -v[v]... for full protocol decode
listening on eth0, link-type EN10MB (Ethernet), snapshot length 262144 bytes

[... no output during invocation ...]
```

**Comment by bzp99** [comment]
Another thing: the method name is actually `initEntries`, not `InitEntries`. This invocation should fail immediately.

**Comment by bzp99** [comment]
The current diagnosis, as discussed on our meeting today plus after some experimenting:

- It does not seem that the CC itself is at fault (I mean the TPC-C port)
- The behaviour has been somehow introduced when changing the build process (→ Java/OpenJML/Gradle versions)
- The same behaviour occurs when trying to compile/execute _without_ OpenJML altogether

Overall, I have a feeling that downgrading to Java 8 is causing some issues. Will try to bump the version to see what happens soon.

**Comment by bzp99** [comment]
Unfortunately, if it really is the Java version, we are at an impasse:

- The latest _stable (?)_ version of OpenJML is [v0.8.59](https://github.com/OpenJML/OpenJML/releases/tag/0.8.59) and it only has support for Java 8
- The _latest latest_ version of OpenJML is an alpha ([0.17.0-alpha-15](https://github.com/OpenJML/OpenJML/releases/tag/0.17.0-alpha-15)) that I remember to be unstable, not to mention that the build process completely changes when using this version

NB v0.8.59 was released on 2021-11-17 and 0.17.0-alpha-15 on 2022-06-06. There have been no developments since then.

**Comment by bzp99** [comment]
I can now confirm that upgrading to Java 17 from Java 8 (OpenJML put aside for now) has fixed the issue. Now we have to deal with the problems outlined in my previous comment.

**Comment by bzp99** [comment]
Resolved by b14b994010a4df5ff9f50be211a2dfbf3485aebd
